refactor: Changes the list views to use the new Select component

michael-s-molina opened the pull request:
### SUMMARY
- Change the list views to use the new Select component
- Adjusts the design of the date range picker to match other controls
- Aligns the sort select with the other controls
- Makes the layout behavior consistent when resizing the screen
- Adjusts the controls to have the same width
- Previously we had 'Me' and 'Admin', which are the same user, now it will only show the name of the logged user always as the first option of the select.
- Adjusts the tests

@junlincc @jinghua-qa @rusackas @mistercrunch 

PS: Sorry for the big PR, but everything needs to be tested as a whole.

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF
https://user-images.githubusercontent.com/70410625/130853912-05e9ac8e-f7d3-483b-9e82-ab629e18cbb4.mov

https://user-images.githubusercontent.com/70410625/133250852-7e282cb6-3113-4012-8393-26257b66cbda.mov

### TESTING INSTRUCTIONS
Check the videos for instructions.

### ADDITIONAL INFORMATION
- [ ] Has associated issue:
- [ ] Required feature flags:
- [x] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

michael-s-molina committed:
chore: Change the list views to use the new Select component

michael-s-molina committed:
Fix Cypress tests

michael-s-molina committed:
Enables search for all controls

michael-s-molina committed:
Adjusts controls width

etr2460 reviewed (commented):
some initial comments

michael-s-molina commented:
@ktmud About the 'Me' behavior. I found it a little strange to have options like 'Me' and 'Jesse Yang' if you're logged in as Jesse Yang. It gives me the impression that we are talking about different users. What do you think about using one or the other? Can I suggest use the user name and remove the 'Me'?

ktmud commented:
Sure, that makes sense. But having "Me" as the first option when users didn't type anything to search is still useful though.

michael-s-molina commented on superset-frontend/src/components/ListView/Filters/index.tsx:
```diff
@@ -34,17 +34,13 @@ interface UIFiltersProps {
 }: UIFiltersProps) {
   return (
-    <FilterWrapper>
+    <>
```
It seems that this feature is not fully supported by Typescript yet. 
https://stackoverflow.com/questions/67908439/react-typescript-how-to-fix-return-type-element-is-not-a-valid-jsx-eleme?noredirect=1&lq=1

michael-s-molina committed:
Removes 'Me' and keeps the logged user on top

michael-s-molina commented on superset-frontend/src/components/ListView/Filters/Select.tsx:
```diff
@@ -75,78 +61,53 @@ function SelectFilter({
+        totalCount: 0,
+      };
+    },
+    [], // eslint-disable-line react-hooks/exhaustive-deps
```
Yes! Done.

michael-s-molina commented:
> 1. When there is no selected value,me is not ranked the first anymore:
> 2. I was not able to find myself while searching my full name:

Previously we had a conflict because we could have at the same time Me and the name of the logged person. That was weird because I could select 'Me' but 'My name' wasn't selected. It also gave the impression that 'Me' and 'My Name' was two different users. To resolve that, I removed the 'Me' concept and now only present the name of the logged user always on top so it's the first option when opening the select.

> 3. Not able to find other filter options when there is a selected value in the URL parameters (to test this, just select a value and refresh the page):

Fixed

> 4. After selecting some values, clicking on the main nav links ("Charts" or "Dashboards") updates the URL but does not reset the filters.

This is also happening on master. I started investigating the problem but I'm afraid we could end up with even more changed files so I will fix this in a separate PR.

michael-s-molina commented on superset-frontend/src/components/ListView/Filters/Select.tsx:
```diff
@@ -75,78 +61,53 @@ function SelectFilter({
-        />
-      )}
+      <Select
+        ariaLabel={typeof Header === 'string' ? Header : emptyLabel}
```
You're right. I changed it to use the header or the name or a generic 'Filter'.

michael-s-molina committed:
Fixes tests

michael-s-molina commented:
/testenv up

geido reviewed (commented):
My only concern with this is the visuals of the filters. I tend to think the lighter design worked better here. The page looks much more crowded now. How do you feel about considering a lighter version of the Select for cases like this one? (no borders, inline header, etc.)

michael-s-molina committed:
Uses the borderless version for the filters

michael-s-molina commented:
> My only concern with this is the visuals of the filters. I tend to think the lighter design worked better here. The page looks much more crowded now. How do you feel about considering a lighter version of the Select for cases like this one? (no borders, inline header, etc.)

@geido This is a great suggestion. It will make the change less impactful for the users. I updated the code and the PR's video.

michael-s-molina committed:
Fixes the tests

pkdotson commented on superset-frontend/src/views/CRUD/chart/ChartList.tsx:
```diff
@@ -417,113 +417,110 @@ function ChartList(props: ChartListProps) {
+    [],
+  );
+
+  const filters: Filters = useMemo(
```
Nice, I think memoize works here.

pkdotson commented:
/testenv up

pkdotson commented on superset-frontend/src/components/ListView/ListView.test.jsx:
```diff
@@ -462,7 +470,7 @@ describe('ListView', () => {
 
     await act(async () => {
-      wrapper2.find('[data-test="card-sort-select"]').first().props().onChange({
+      wrapper2.find('[aria-label="Sort"]').first().props().onChange({
```
nit: should this be lower case?

michael-s-molina replied:
We are now referring to the `aria-label` property, which doesn't have to be lowercase.

pkdotson reviewed (approved):
Code looks good to me. Minor the other suggestions, nits, and design review I'll conditionally approve.

michael-s-molina committed:
Reverts the Select theme to the default

michael-s-molina commented:
After a session with the design team, we decided to keep the default style of the Select component to preserve consistency across the product. I updated the PR's video.

ktmud commented:
/testenv up

ktmud commented:
![image](https://user-images.githubusercontent.com/335541/133335186-21268e3e-7052-4e4f-a2fd-eeeec0fc6f74.png)

I'm getting a JS error when trying to clear a selected value.

michael-s-molina committed:
Rebases and fixes js error

michael-s-molina commented:
> I'm getting a JS error when trying to clear a selected value.

Thanks for spotting that @ktmud. I fixed and also rebased to get the latest changes.

michael-s-molina committed:
Fixes failing test

michael-s-molina commented:
/testenv up

ktmud commented on superset-frontend/src/components/ListView/Filters/Select.tsx:
```diff
@@ -16,137 +16,76 @@
-  PartialThemeConfig,
-} from 'src/components/Select';
+import React, { useState, useMemo } from 'react';
+import { withTheme, SupersetThemeProps, t } from '@superset-ui/core';
```
It seems theme variables are not used directly by this component anymore. Can we clean up `withTheme`?

On a separate note, I think we should prefer `useTheme` in all future refactoring, too---unless for wrapping legacy class-based components. I added an entry in the [CSS Styled Guide](https://github.com/apache/superset/wiki/Emotion-Styling-Guidelines-and-Best-Practices/_compare/4ea6764f497f352bff4742b3dfa192b4c46d633a...9cd7da713051b2b98eeb512fd92da2b707145c02#diff-5d6646c4ad4792858d153e144bf5fac28133c6005ed3bfd54ccaddde9c22732eR11).

cc. @evans @etr2460 Lmk if you have different opinions.

michael-s-molina replied:
Agreed

michael-s-molina replied:
`withTheme` removed.

ktmud reviewed (approved):
Code lgtm but are we sure this is the design we want? Currently it jus looks like there are too many "holes" in the header section. Consistency is good for most cases, but there is also context here---IMO a simple dropdown switch used for navigation has a totally different purpose than say a multi-select control in a form intended for collecting information.

This is deviating further and further away from SIP-34 design:

<img width="1427" alt="Xnip2021-09-15_17-42-36" src="https://user-images.githubusercontent.com/335541/133650760-7ff71087-0bf4-4586-a9cf-41b6fc6042e6.png">

I'm going to stamp as I understand the need to consolidate Select implementation, but I hope we can reconsider the design here.

michael-s-molina commented:
> This is deviating further and further away from SIP-34 design:
> 
> I'm going to stamp as I understand the need to consolidate Select implementation, but I hope we can reconsider the design here.

@ktmud During our design session we came to the same conclusion. We chose to preserve the Select consistency for now, in order to migrate the component, but @mihir174 and @jess-dillard will analyze the design and propose improvements that will be implemented in the future.

michael-s-molina commented on superset-frontend/src/components/DatePicker/index.tsx:
```diff
@@ -17,6 +17,12 @@
+
+export const RangePicker = styled(AntdRangePicker)`
+  border-radius: ${({ theme }) => theme.gridUnit}px;
+`;
```
Yes. The idea is to match the border style of the `Select` and `Input` components.

ktmud commented:
<img width="1276" alt="Xnip2021-09-16_11-29-47" src="https://user-images.githubusercontent.com/335541/133665762-6052f897-8016-42de-950b-fef34d7970a0.png">

Not sure if it'd help but here's a design I implemented for another project. Using carets instead of arrows helps.

michael-s-molina committed:
Removes unused withTheme
